src/r46ca01.toit

```diff
@@ -0,0 +1,158 @@
+// Copyright (C) 2022 Toitware ApS. All rights reserved.
+// Use of this source code is governed by a MIT-style license that can be found
+// in the LICENSE file.
+
+import modbus
+import modbus.rs485 as modbus
+import rs485
+import gpio
+
+/**
+A driver for the R46CA01 temperature sensor.
+*/
+class R46ca01:
+  static DEFAULT_UNIT_ID ::= 1
+  static DEFAULT_BAUD_RATE ::= 9600
+
+  static TEMPERATURE_ADDRESS_ ::= 0x00
+  static UNIT_ID_ADDRESS_ ::= 0x02
+  static BAUD_RATE_ADDRESS_ ::= 0x03
+  static CORRECTION_ADDRESS_ ::= 0x04
+
+  static BAUD_RATE_1200_ ::= 0
+  static BAUD_RATE_2400_ ::= 1
+  static BAUD_RATE_4800_ ::= 2
+  static BAUD_RATE_9600_ ::= 3
+  static BAUD_RATE_19200_ ::= 4
+
+  registers_/modbus.HoldingRegisters
+
+  /**
+  Creates a new R46CA01 driver.
+
+  The given Modbus $station must be a R46CA01 device.
+  */
+  constructor station/modbus.Station:
+    registers_ = station.holding_registers
+
+  /**
+  Creates a new R46CA01 driver.
+
+  Uses $detect_unit_id to find the unit id of the R46CA01 device.
+  The R46CA01 device must be the only device on the bus.
+  */
+  constructor.detect bus/modbus.Modbus:
+    id := detect_unit_id bus
+    return R46ca01 (bus.station id)
+
+  /**
+  Reads the unit id (also known as "server address", or "station address") from the connected sensor.
+
+  Note that only one unit must be on the bus when performing this action.
+  */
+  static detect_unit_id bus/modbus.Modbus -> int:
+    broadcast_station := bus.station 0xFF
+    return broadcast_station.holding_registers.read_single --address=UNIT_ID_ADDRESS_
+
+  /**
+  Reads the temperature.
+
+  Returns the result in degrees Celsius.
+  */
+  read_temperature -> float:
+    raw := read_temperature --raw
+    if raw == 0x8000: throw "NO_DS18B20_OR_ERROR"
+    return raw * 0.1
+
+  /**
+  Reads the temperature and returns the raw value.
+
+  Returns the value as given by the sensor.
+  If the returned value equals 0x8000 then no DS18B20 is connected to the
+    board, or another error was encountered.
+
+  Each unit corresponds to 0.1 degrees Celsius.
+  */
+  read_temperature --raw/bool -> int:
+    if not raw: throw "INVALID_ARGUMENT"
+    return read_ TEMPERATURE_ADDRESS_
+
+  /**
+  Reads the correction value.
+
+  Returns the correction value in degrees Celsius.
+  */
+  read_correction -> float:
+    return (read_correction --raw) * 0.1
+
+  /**
+  Reads the correction value.
+
+  Returns the raw correction value as reported by the sensor. Each unit corresponds
+    to 0.1 degrees Celsius.
+  */
+  read_correction --raw/bool -> int:
+    if not raw: throw "INVALID_ARGUMENT"
+    return read_ CORRECTION_ADDRESS_
+
+  /**
+  Sets the correction value.
+
+  The reported temperature is adjusted by this value. A positive value
+    increases the temperature, and a negative value decreases it.
```

Comment:
Is it stored non-volatile, or do you have to do this every time?

Reply:
non-volatile.
I updated the comment.
done.